cmd/thanos/tools_stream_metric.go

```diff
@@ -0,0 +1,167 @@
+}
+
+func streamMetric(conf *rawMetricConfig, logger log.Logger) error {
+	nowTime := time.Now()
+	startTime := nowTime.Add(-time.Duration(conf.hoursAgo) * time.Hour)
+	endTime := startTime.Add(time.Duration(conf.hours) * time.Hour)
+	conn, err := grpc.Dial(
+		conf.storeAddr,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	if err != nil {
+		level.Error(logger).Log("msg", "Failed to create gRPC client", "err", err, "store", conf.storeAddr)
+		return err
+	}
+	storeClient := storepb.NewStoreClient(conn)
+	labelMatchers := []storepb.LabelMatcher{
+		{Type: storepb.LabelMatcher_EQ, Name: "__name__", Value: conf.metric},
+	}
+	addMatcher := func(mtype storepb.LabelMatcher_Type, matcher string) {
+		parts := strings.Split(matcher, ":")
+		if len(parts) != 2 {
+			level.Error(logger).Log("msg", "ignoring an invalid label matcher", "matcher", matcher)
+			return
+		}
+		labelMatchers = append(labelMatchers, storepb.LabelMatcher{
+			Type:  mtype,
+			Name:  parts[0],
+			Value: parts[1],
+		})
+	}
+	addMatcher(storepb.LabelMatcher_EQ, conf.equalLabelMatcher)
+	addMatcher(storepb.LabelMatcher_NEQ, conf.notEqualLabelMatcher)
+	storeReq := &storepb.SeriesRequest{
+		Aggregates: []storepb.Aggr{storepb.Aggr_RAW},
+		Matchers:   labelMatchers,
+		MinTime:    startTime.UnixMilli(),
+		MaxTime:    endTime.UnixMilli(),
+		SkipChunks: conf.skipChunks,
+	}
+
+	level.Info(logger).Log(
+		"msg", "sending a gRPC call to store",
+		"num_label_matchers", len(storeReq.Matchers),
+		"label_matchers", fmt.Sprintf("%+v", storeReq.Matchers),
+		"req", storeReq.String(),
+	)
+
+	storeRes, err := storeClient.Series(context.Background(), storeReq)
+	if err != nil {
+		return err
+	}
+	seq := 0
+	for ; true; seq++ {
+		resp, err := storeRes.Recv()
+		if err == io.EOF {
+			level.Info(logger).Log("msg", "Got EOF from store")
+			break
+		}
+		if err != nil {
+			return err
+		}
+		warning := resp.GetWarning()
+		if len(warning) > 0 {
+			level.Warn(logger).Log("msg", "got a warning from store", "warning", warning)
+			if !conf.ignoreWarnings {
+				return fmt.Errorf("got a warning from store: %v", warning)
+			}
+		}
+		series := resp.GetSeries()
+		if series == nil {
+			return fmt.Errorf("unexpected nil series")
+		}
+		if (seq % 1000) == 0 {
+			level.Info(logger).Log("msg", "streaming time series", "seq", seq)
+		}
+		fmt.Printf("%s{", conf.metric)
+		comma := ""
+		for _, label := range series.Labels {
+			if label.Name != labels.MetricName {
+				fmt.Printf("%s%s=\"%s\",", comma, label.Name, label.Value)
+				comma = ","
+			}
+		}
+		fmt.Printf("}")
+		for i, chunk := range series.Chunks {
+			raw, err := chunkenc.FromData(chunkenc.EncXOR, chunk.Raw.Data)
+			if err != nil {
+				level.Error(logger).Log("err", err, "msg", "error in decoding chunk")
+				return err
+			}
+			fmt.Printf("  chunk %d:", i)
+			iter := raw.Iterator(nil)
+			for iter.Next() != chunkenc.ValNone {
+				ts, value := iter.At()
+				fmt.Printf(" [%d, %.2f]", ts, value)
+			}
```

Comment:
Could probably split this out into some expandChunk method. Also, maybe we want to print out histograms too?

Reply:
The intention here is to print raw data samples in a format that is compatible with other tools (e.g., Python), so they can be used for further analysis. I believe raw data samples are sufficient for this purpose.